### PR TITLE
Inline statically known method errors.

### PR DESCRIPTION
**Pull request by gbaraldi**
This replaces the `Expr(:call, ...)` with a call of a new builtin `Core.throw_methoderror`

This is useful because it makes very clear if something is a static method error or a plain dynamic dispatch that always errors. 
Tools such as AllocCheck or juliac can notice that this is not a genuine dynamic dispatch, and prevent it from becoming a false positive compile-time error.

Dependent on https://github.com/JuliaLang/julia/pull/55705

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
I generally dislike optimizing codepaths that aren't performance critical, as it bloats the code, slows the compiler, and introduces more possibilities for subtle discrepancies in what error is generated and thrown

**Comment by topolarity** [comment]
> I generally dislike optimizing codepaths that aren't performance critical, as it bloats the code, slows the compiler, and introduces more possibilities for subtle discrepancies in what error is generated and thrown

I share that concern, but in this case the transformation is pretty minimal I think.

It turns:
```julia
Expr(:call, foo, args...)::Union{}
```
into:
```julia
%argtuple = Expr(:call, tuple, args...)
%err = Expr(:new, MethodError, farg, %argtuple, %world)
Expr(:call, throw, %err)
```

(note that since https://github.com/JuliaLang/julia/pull/54537 the `%world` computation is lifted to function entry)

Those 2 extra instructions don't seem too bad to me.

File: base/compiler/ssair/verify.jl [hunk 1]
```diff
@@ -1,9 +1,29 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 function maybe_show_ir(ir::IRCode)
-    if isdefined(Core, :Main)
+    if isdefined(Core, :Main) && isdefined(Core.Main, :Base)
         # ensure we use I/O that does not yield, as this gets called during compilation
         invokelatest(Core.Main.Base.show, Core.stdout, "text/plain", ir)
+    else
+        print(Core.stderr, "Verification error in IR for function ")
```
Comment (topolarity) [inline base/compiler/ssair/verify.jl:8]: I think these changes slipped in by accident

File: base/compiler/ssair/inlining.jl [hunk 2]
```diff
@@ -74,6 +75,16 @@ function add_inlining_backedge!((; edges, invokesig)::InliningEdgeTracker, mi::M
     return nothing
 end
 
+function add_inlining_mt_backedge!((; edges, invokesig)::InliningEdgeTracker, mt::MethodTable)
```
Comment (topolarity) [inline base/compiler/ssair/inlining.jl:78]: I think this ended up unused

File: base/compiler/abstractinterpretation.jl [hunk 3]
```diff
@@ -511,9 +510,10 @@ function add_call_backedges!(interp::AbstractInterpreter, @nospecialize(rettype)
     # also need an edge to the method table in case something gets
     # added that did not intersect with any existing method
     if isa(matches, MethodMatches)
-        matches.fullmatch || add_mt_backedge!(sv, matches.mt, atype)
+        matches.fullmatch || add_mt_backedge!(sv, matches.info.mt, atype)
     else
-        for (thisfullmatch, mt) in zip(matches.fullmatches, matches.mts)
+        matches::UnionSplitMethodMatches
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:515]:
```suggestion
```

File: base/compiler/ssair/inlining.jl [hunk 4]
```diff
@@ -1496,19 +1538,19 @@ function concrete_result_item(result::ConcreteResult, @nospecialize(info::CallIn
 end
 
 function handle_cases!(todo::Vector{Pair{Int,Any}}, ir::IRCode, idx::Int, stmt::Expr,
-    @nospecialize(atype), cases::Vector{InliningCase}, all_covered::Bool,
+    @nospecialize(atype), cases::Vector{InliningCase}, handled_all_cases, fully_covered,
```
Comment (aviatesk) [inline base/compiler/ssair/inlining.jl:1541]:
```suggestion
    @nospecialize(atype), cases::Vector{InliningCase}, handled_all_cases::Bool, fully_covered::Bool,
```

File: base/compiler/ssair/inlining.jl [hunk 5]
```diff
@@ -1392,22 +1421,35 @@ function compute_inlining_cases(@nospecialize(info::CallInfo), flag::UInt32, sig
...
+            if unwrapped_info isa UnionSplitInfo
+                for mt in unwrapped_info.mts
+                    push!(state.edges, mt, atype)
+                end
+            elseif unwrapped_info isa MethodMatchInfo
+                push!(state.edges, unwrapped_info.mt, atype)
+            else
+                @assert false
+            end
+        end
```
Comment (aviatesk) [inline base/compiler/ssair/inlining.jl:1437]: I don't think it's necessary to add backedges for all tables.
I might be a better design to set up a new inlining case, something like `MethodErrorCase` and append it to `cases` during the iteration of matches while adding only the necessary backedges.

Comment (aviatesk) [reply base/compiler/ssair/inlining.jl:1437]: And it looks like we now need a new `CallInfo` interface to lookup a method table from any call info object.

**Comment by aviatesk** [comment]
This change does not seem to bloat up the sysimg, but I have the same concern and would prefer to make it an optional feature enabled by a new optimization parameter.

File: base/compiler/ssair/inlining.jl [hunk 6]
```diff
@@ -597,18 +608,40 @@ function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int, argexprs::
...
+
+Insert a static method error for the call expression at `idx` in `compact`
+The call is inserted at the current IncrementalCompact cursor
+"""
+function ir_inline_method_error!(compact::IncrementalCompact, idx::Int, argexprs::Vector{Any}, interp::AbstractInterpreter)
+    line = compact.result[idx][:line]
+    world_age_stmt = Expr(:foreigncall, :(:jl_get_tls_world_age), UInt, svec(), 0, :(:ccall))
+    flags = recompute_effects_flags(optimizer_lattice(interp), world_age_stmt, UInt, compact)
+    world_age = insert_node_here!(compact, NewInstruction(world_age_stmt, UInt, NoCallInfo(), line, flags))
+    argtuple = insert_node_here!(compact, NewInstruction(Expr(:call, tuple, argexprs[2:end]...), Tuple, line))
```
Comment (aviatesk) [inline base/compiler/ssair/inlining.jl:637]:
```suggestion
    argtuple = insert_node_here!(compact, NewInstruction(Expr(:call, GlobalRef(Core, :tuple), argexprs[2:end]...), Tuple, line))
    world_age = insert_node_here!(compact, NewInstruction(world_age_stmt, UInt, NoCallInfo(), line, flags))
```

File: base/compiler/ssair/inlining.jl [hunk 7]
```diff
@@ -597,18 +608,40 @@ function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int, argexprs::
...
+
+    methoderr_stmt = Expr(:new, MethodError, argexprs[1], argtuple, world_age)
+    methoderr = insert_node_here!(compact, NewInstruction(methoderr_stmt, MethodError, line))
+    insert_node_here!(compact, NewInstruction(Expr(:call, throw, methoderr), Union{}, line))
```
Comment (aviatesk) [inline base/compiler/ssair/inlining.jl:641]:
```suggestion
    insert_node_here!(compact, NewInstruction(Expr(:call, GlobalRef(Core, :throw), methoderr), Union{}, line))
```

File: base/compiler/ssair/inlining.jl (hunk 6 above)
Comment (topolarity) [reply base/compiler/ssair/inlining.jl:637]: Question out of personal ignorance: is there any particular reason for the swap?

**Comment by topolarity** [comment]
> This change does not seem to bloat up the sysimg, but I have the same concern and would prefer to make it an optional feature enabled by a new optimization parameter.

We can do that, but it puts a tool like AllocCheck in an awkward position since it would now want to run with custom inference for precision purposes - but that can now diverge from the codegen/analysis for the standard inference settings (even if the dynamic dispatch itself is harmless in both cases) so `@check_allocs` in your tests may not give you the guarantees you want at runtime when you don't use it

That guarantee is already weak (since inference can be noticeably stateful in, e.g., recursive cycles), but I'd prefer not to make it worse if we can avoid it

File: base/compiler/ssair/inlining.jl (hunk 5 above)
Comment (topolarity) [reply base/compiler/ssair/inlining.jl:1437]: > I don't think it's necessary to add backedges for all tables.

That's true - To do that right, I think we could move `UnionSplitMethodMatches.fullmatches` into `UnionSplitInfo` and check that here

Comment (topolarity) [reply base/compiler/ssair/inlining.jl:1437]: The case would be when we are fully covering within one MethodTable but not the other - in that case we only need the backedge to the non-fully-covering MethodTable

See the computation of `fullmatches` here: https://github.com/JuliaLang/julia/blob/00c700ee58f3393c0b733497cd2520129a0907c6/base/compiler/abstractinterpretation.jl#L299

File: base/compiler/ssair/inlining.jl (hunk 7 above)
Comment (topolarity) [reply base/compiler/ssair/inlining.jl]: I made this change, but fwiw there are several places in this pass where we don't do this

Is it important?

**Comment by topolarity** [comment]
I'm very surprised by the CI failure:
> throw=true: Test Failed at C:\buildkite-agent\builds\win2k22-amdci6-0\julialang\julia-master\julia-6cebcee53d\share\julia\test\threads_exec.jl:1237
--
  | Expression: begin
  | #= C:\buildkite-agent\builds\win2k22-amdci6-0\julialang\julia-master\julia-6cebcee53d\share\julia\test\threads_exec.jl:1238 =#
  | waitany(convert_tasks(tasks_type, tasks); throw = true)
  | end
  | Expected: CompositeException
  | No exception thrown

https://buildkite.com/julialang/julia-master/builds/37824#01906ea8-a916-4658-938a-10b4b9e027f2/665-1154

**Comment by aviatesk** [comment]
> To avoid IR bloat, would it preferable to lower this to a new apply-like builtin instead?
> 
> Something like `Expr(:call, Core.throw_method_error, f, args...)` or `Expr(:method_error, ...)`

This seems like one possible approach, but adding it solely for this purpose feels a bit hesitant. This is because inlining the code to raise a `MethodError` in this case offers (almost) no performance or latency benefits. That said, we do add new intrinsics quite frequently, so it might not be a big issue though.
Instead, I believe it would be cleaner to leave the fallback `:call` in such cases, but also set its `:type` to `Union{}`, and make the next statement `:unreachable`? The ultimate goal of this PR is, if I understand correctly, to make the "must-throw" information for such `:call` visible at the LLVM level. And inserting `:unreachable` seems to be the cleanest solution to achieve that.

**Comment by aviatesk** [comment]
I see. But if we assume that allocations on throw code path are not performance-critical, wouldn't it be okay for AllocCheck to ignore such calls?
Since the current implementation of inference is based on that assumption, it seems natural to reflect that in AllocCheck as well.

**Comment by topolarity** [comment]
AllocCheck also supports analysis including throw paths, it's just not turned on by default.

File: test/threads_exec.jl [hunk 8]
```diff
@@ -1233,6 +1233,7 @@ end
             @testset "throw=true" begin
                 tasks, event = create_tasks()
                 push!(tasks, Threads.@spawn error("Error"))
+                wait(tasks[end]; throw=false)
```
Comment (Keno) [inline test/threads_exec.jl:1236]: Unrelated?

File: base/compiler/ssair/inlining.jl [hunk 9]
```diff
@@ -597,18 +598,40 @@ function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int, argexprs::
...
     return insert_node_here!(compact, NewInstruction(pn, typ, line))
 end
 
+"""
+    ir_inline_method_error!
+
+Insert a static method error for the call expression at `idx` in `compact`
+The call is inserted at the current IncrementalCompact cursor
+"""
+function ir_inline_method_error!(compact::IncrementalCompact, idx::Int, argexprs::Vector{Any}, interp::AbstractInterpreter)
```
Comment (Keno) [inline base/compiler/ssair/inlining.jl:622]: I think for IR size, this would be better off as a `throw_methoderror` builtin that just takes the arguments.

File: test/threads_exec.jl (hunk 8 above)
Comment (topolarity) [reply test/threads_exec.jl:1236]: This was a race condition in the test that started failing after these changes, for some reason

**Comment by topolarity** [comment]
Alright this now turns the call into:
```julia
Core.throw_methoderror(f, args...)
unreachable
```

which should hopefully resolve the IR bloat concerns

File: base/compiler/ssair/inlining.jl [hunk 10]
```diff
@@ -244,7 +245,7 @@ function cfg_inline_unionsplit!(ir::IRCode, idx::Int, union_split::UnionSplit,
         end
     end
     # The edge from the fallback block.
-    fully_covered || push!(from_bbs, length(state.new_cfg_blocks))
+    !handled_all_cases && push!(from_bbs, length(state.new_cfg_blocks))
```
Comment (aviatesk) [inline base/compiler/ssair/inlining.jl:248]:
```suggestion
    # NOTE In the cases of `!fully_covered`, since no additional edge is required since
    #      we will inline `Core.throw_methoderror` for this dispatch candidate, which is 
    #      must-throw, making the subsequent code path unreachable.
    !handled_all_cases && push!(from_bbs, length(state.new_cfg_blocks))
```

File: base/compiler/ssair/inlining.jl [hunk 11]
```diff
@@ -1392,22 +1396,33 @@ function compute_inlining_cases(@nospecialize(info::CallInfo), flag::UInt32, sig
...
+            atype = argtypes_to_type(sig.argtypes)
+            # We will emit an inline MethodError so we need a backedge to the MethodTable
+            unwrapped_info = info isa ConstCallInfo ? info.call : info
+            if unwrapped_info isa UnionSplitInfo
+                for (fullmatch, mt) in zip(unwrapped_info.fullmatches, unwrapped_info.mts)
+                    !fullmatch && push!(state.edges, mt, atype)
+                end
+            elseif unwrapped_info isa MethodMatchInfo
+                push!(state.edges, unwrapped_info.mt, atype)
+            else @assert false end
```
Comment (aviatesk) [inline base/compiler/ssair/inlining.jl:1409]: This assertion does not hold in cases where external `CallInfo` is used, e.g. the test case in the following diff causes this code path to be hit:
```diff
diff --git a/test/compiler/AbstractInterpreter.jl b/test/compiler/AbstractInterpreter.jl
index d95354cefa..6c745218c0 100644
--- a/test/compiler/AbstractInterpreter.jl
+++ b/test/compiler/AbstractInterpreter.jl
@@ -431,6 +431,8 @@ end
 @inline function inlined_usually(x, y, z)
     return x * y + z
 end
+foo_split(x::Float64) = 1
+foo_split(x::Int) = 2
 
 # check if the inlining algorithm works as expected
 let src = code_typed1((Float64,Float64,Float64)) do x, y, z
@@ -444,6 +446,7 @@ let NoinlineModule = Module()
     main_func(x, y, z) = inlined_usually(x, y, z)
     @eval NoinlineModule noinline_func(x, y, z) = $inlined_usually(x, y, z)
     @eval OtherModule other_func(x, y, z) = $inlined_usually(x, y, z)
+    @eval NoinlineModule bar_split_error() = $foo_split(Core.compilerbarrier(:type, nothing))
 
     interp = NoinlineInterpreter(Set((NoinlineModule,)))
 
@@ -473,6 +476,11 @@ let NoinlineModule = Module()
         @test count(isinvoke(:inlined_usually), src.code) == 0
         @test count(iscall((src, inlined_usually)), src.code) == 0
     end
+
+    let src = code_typed1(NoinlineModule.bar_split_error)
+        @test count(iscall((src, foo_split)), src.code) == 0
+        @test count(iscall((src, Core.throw_methoderror)), src.code) > 0
+    end
 end
 
 # Make sure that Core.Compiler has enough NamedTuple infrastructure
```

To handle such cases, we probably need to introduce a new `CallInfo` interface `getmatchinfo(::CallInfo, idx::Int)::Tuple{#=fullmatch=#Bool, #=mt=#MethodTable}` and enforce its implementation for each `CallInfo`, and use it here:
```suggestion
            for i = 1:nunion
                fullmatch, mt = getmatchinfo(info, i)
                !fullmatch && push!(state.edges, mt, atype)
            end
```

**Comment by aviatesk** [comment]
LGTM modulo the comment posted above.

File: base/compiler/ssair/inlining.jl (hunk 11 above)
Comment (topolarity) [reply base/compiler/ssair/inlining.jl:1409]: Oh, that's a very good catch (and a useful test!) thanks.

That change looks pretty easy - I'll add it to the PR.

Still, I'm curious - It seems like `abstract_applicable` / `find_matching_methods` do not support custom `CallInfo`, etc. Is that because those are modeling a function application via the global method table, and therefore we expect the "standard" CallInfo semantics?

**Comment by gbaraldi** [comment]
Should we run benchmarks/pkgeval?

**Comment by topolarity** [comment]
@nanosoldier runbenchmarks(vs=":master")

PkgEval is probably unnecessary - this shouldn't have any observable semantic differences

**Comment by topolarity** [comment]
@nanosoldier `runbenchmarks(vs=":master")`

PkgEval is probably unnecessary - this shouldn't have any observable semantic differences

**Comment by topolarity** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

PkgEval is probably unnecessary - this shouldn't have any observable semantic differences

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/54972#issuecomment-2341521799) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/86e3a38_vs_bcf41ba/report.md).

File: base/compiler/ssair/inlining.jl (hunk 11 above)
Comment (aviatesk) [reply base/compiler/ssair/inlining.jl]: > Still, I'm curious - It seems like `abstract_applicable` / `find_matching_methods` do not support custom `CallInfo`, etc. Is that because those are modeling a function application via the global method table, and therefore we expect the "standard" CallInfo semantics?

In fact, `find_matching_methods` is meant to retrieve the results of method matching, and it is not intended to create `CallInfo`, especially `MethodMatchInfo` or `UnionSplitInfo`. Currently, these objects are created within `find_matching_methods` for convenience, but I guess it might be more sensible for the caller side, such as `abstract_call_gf_by_type` (which calls `find_matching_methods`), to create those `CallInfo` objects. Custom `CallInfo` is used when an external `AbstractInterpreter` injects some logic into those routines, typically to trigger special behavior such as inlining.

File: base/compiler/types.jl [hunk 12]
```diff
@@ -451,9 +451,12 @@ abstract type CallInfo end
 nsplit(info::CallInfo) = nsplit_impl(info)::Union{Nothing,Int}
 getsplit(info::CallInfo, idx::Int) = getsplit_impl(info, idx)::MethodLookupResult
 getresult(info::CallInfo, idx::Int) = getresult_impl(info, idx)
+add_uncovered_edges!(edges::Vector{Any}, info::CallInfo, @nospecialize(atype)) = add_uncovered_edges_impl(edges, info, atype)
+
 
 nsplit_impl(::CallInfo) = nothing
 getsplit_impl(::CallInfo, ::Int) = error("unexpected call into `getsplit`")
 getresult_impl(::CallInfo, ::Int) = nothing
+add_uncovered_edges_impl(edges::Vector{Any}, info::CallInfo, @nospecialize(atype)) = nothing
```
Comment (aviatesk) [inline base/compiler/types.jl:460]: With this default implementation, for any arbitrary custom `CallInfo`, as long as it correctly implements `getsplit_impl`, the method error case inlining enabled by this PR will occur. However, if that `CallInfo` does not implement `add_uncovered_edges_impl`, there could theoretically be cases where the necessary edges are not added. In the version I proposed, the `getmatchinfo(::CallInfo, ::Int)` interface is enforced to be implemented for all `CallInfo`s, so I think such interface requirement errors are easy to detect. That said, this is a very rare case, so I think it’s just fine to revisit it later when there’s more time.

**Review by aviatesk — APPROVED** [review]

File: base/compiler/types.jl (hunk 12 above)
Comment (aviatesk) [reply base/compiler/types.jl:460]: > The `getmatchinfo(::CallInfo, ::Int)` isn't quite right, because it treats uncovered edges as being associated with split indices, but they are the edge(s) not covered by the split cases (e.g. you might have 10 split cases, and then an uncovered `atype` that needs an edge to 2 method tables)

Right... So maybe `getuncoveredmt(::CallInfo) -> iterator of method tables with !fullmatch`?

Comment (aviatesk) [reply base/compiler/types.jl:460]: > why we provide a default implementation for any of these

Since they are optional. If `CallInfo` implements `nsplit`, then the inlining for the call is enabled and the `CallInfo` is required to implement `getsplit`. And `getresult` is optional, and if it's implemented, then the inliner may use a result from extended lattice inference (a.k.a. const-prop and semi-concrete interpretation).

Comment (topolarity) [reply base/compiler/types.jl:460]: That's helpful - sounds like this one should be:
```julia
add_uncovered_edges_impl(edges::Vector{Any}, info::CallInfo, @nospecialize(atype)) = error("unexpected call into add_uncovered_edges!")
```

since we also only do this if you implement `nsplit` 

That said, `getuncoveredmt` would also be OK w/ me - the API shape seems orthogonal to the optionality, etc.

Comment (vtjnash) [reply base/compiler/types.jl]: I intend for us to eliminate all this tracking code for better correctness, as soon as someone can help finish https://github.com/JuliaLang/julia/pull/54894

File: base/compiler/stmtinfo.jl [hunk 13]
```diff
@@ -48,20 +51,27 @@ each partition (`info.matches::Vector{MethodMatchInfo}`).
 This info is illegal on any statement that is not a call to a generic function.
 """
 struct UnionSplitInfo <: CallInfo
-    matches::Vector{MethodMatchInfo}
+    matches::Vector{MethodLookupResult}
+    mts::Vector{MethodTable}
+    fullmatches::Vector{Bool}
```
Comment (aviatesk) [inline base/compiler/stmtinfo.jl:56]: Would it be okay to revert back to storing `Vector{MethodMatchInfo}` like before? The current design, where `matches`/`mts`/`fullmatches` can have different lengths, feels a bit tricky to handle. I understand it's meant to avoid adding duplicate edges to the same `mt` in `add_uncovered_edges_impl`, but in practice, adding duplicate edges isn't really an issue. From a complexity standpoint, I think the previous approach is preferable.

Comment (topolarity) [reply base/compiler/stmtinfo.jl:56]: How would you feel about:
```suggestion
    matches::Vector{MethodLookupResult}
    mt_edges::Vector{@NamedTuple{mt::MethodTable, fullmatch::Bool}}
```

or similar?

That's more consistent with what `UnionSplitMethodMatches` carried before, which always had these edges de-duplicated.

Comment (aviatesk) [reply base/compiler/stmtinfo.jl:56]: Right, the previous design of `UnionSplitMethodMatches` also avoided adding duplicated edges. But I still feel that it's better for `UnionSplitInfo` or `UnionSplitMethodMatches` to have a simpler data structure. We can avoid duplication when adding edges.

Comment (topolarity) [reply base/compiler/stmtinfo.jl:56]: Sure, feel free to open a PR
